### PR TITLE
[CM H2] created interactive tutorial in google codelab for writing a …

### PR DESCRIPTION
**Pull request by aaronabebe**
…specification with elektra. 

This relates to #1522.

I made a new folder in the tutorials directory (`/doc/tutorials/interactive/codelab`), but since this is a new type of tutorial maybe a new folder would be better? 
I also was not really sure what files you would need in the repo to host the tutorial. 
Right now I added 
- the markdown file to generate the html with the google codelab cmd tool `claat`
- the index.html if the tutorial should be served directly as is

I also hosted the current version of the tutorial in this PR at [https://aaronabebe.gitlab.io/](https://aaronabebe.gitlab.io/#0), so that it's a bit easier to look over it.

I was thinking maybe it could also be a good idea to design a landing page for all interactive tutorials in the style of the one from [Google](https://codelabs.developers.google.com/).

## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

**Comment by markus2330** [comment]
Unfortunately, I get a 404 on https://aaronabebe.gitlab.io/

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+
+### Step 1: Mount a specification file
+
+First you need to mount a specification file, in this case `spec.ni` to the `spec:/` namespace.
+We define the path inside the `spec:/` namespace as `/sw/app/current/#0`, this is an arbitrary name and you can choose whatever you want here.
+
+We also need the specify the plugin we use for writing to the file in the correct format. In this case we choose the `ni` plugin to write to the specification file.
+
+``` bash
+kdb mount ~/spec.ni spec:/sw/app/current/#0 ni
```
Comment (markus2330) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:74]:
```suggestion
sudo kdb mount `pwd`/spec.ni spec:/sw/app/current/#0 ni
```

mount should use sudo

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+ =
+
+[]
+ meta:/mountpoint = app.ni
+```
+
+### Step 3: Do a specification mount
+
+```bash
+kdb spec-mount /sw/app/current/#0 ni
```
Comment (markus2330) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:103]:
```suggestion
sudo kdb spec-mount /sw/app/current/#0 ni
```

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+
+
+### Step 2: Adding more metadata
+
+So with your first key added, we of course want to specify more information for the port. There surely is more information to a port than just the type. 
+What about a `default`, or what about an `example` for a usable port? Maybe a `description` what the port really is for?
+Let's add that next!
+
+```bash 
+kdb meta-set spec:/sw/app/current/#0/server/port default 8080
```
Comment (markus2330) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:150]: Make sure to also fix all these key names

**Review by markus2330 — COMMENTED** [review]
Very nicely and freshly written! Unfortunately, I could not see the result.

**Comment by markus2330** [comment]
Btw. would be good if you can integrate https://github.com/ElektraInitiative/libelektra/blob/5702ce0496823fdd12c4e1e6ff9d6d1ca1b71a40/tests/shell/shell_recorder/tutorial_wrapper/README.md so that the build server verifies that the tutorial still works.

File: doc/tutorials/interactive/codelabs/elektra-specification.md (hunk 1 above)
Comment (kodebach) [reply doc/tutorials/interactive/codelabs/elektra-specification.md:74]: Neither `~/spec.ni` nor <code>&#96;pwd&#96;/spec.ni</code> will work as expected. Because you didn't specify a resolver, the default resolver will be used. In a standard setup this means the path will be seen as relative to `/usr/share/elektra/specification/`.

You should use one of these (the first one is the preferred solution)
- `sudo kdb mount spec.ni spec:/sw/app/current/#0 ni` (possibly with a more unique name instead of `spec.ni`)
- `sudo kdb mount --resolver=noresolver "$PWD/spec.ni" spec:/sw/app/current/#0 ni`

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+So we say the concrete configuraion should be written to `app.ni`.
+
+```bash 
+kdb meta-set spec:/sw/app/current/#0 mountpoint app.ni
+```
+
+Your `spec.ni` file should now look something like this:
+
+```bash 
+cat ~/spec.ni
```
Comment (kodebach) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:89]:
```suggestion
cat $(kdb file spec:/sw/app/current/#0)
```

This way it is guaranteed that we use the same file as Elektra itself.

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+``` bash 
+kdb meta-set spec:/sw/app/current/#0/server/port type short
+```
+
+What you also specified in the command above is the type of the configuration key. Elektra uses the [CORBA type system](https://www.libelektra.org/plugins/type)
+and will check that keys conform to the type specified. 
+
+So after adding the initial key your specification should look something like this: 
+```
+cat ~/spec.ni
```
Comment (kodebach) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:126]:
```suggestion
cat $(kdb file spec:/sw/app/current/#0)
```

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+
+```bash
+kdb spec-mount /sw/app/current/#0 ni
+```
+
+
+Your final specification after adding the port should now look something like this
+
+```bash 
+cat ~/spec.ni
```
Comment (kodebach) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:178]:
```suggestion
cat $(kdb file spec:/sw/app/current/#0)
```

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+kdb meta-set spec:/sw/app/current/#0/database/ip default 127.0.0.1
+kdb meta-set spec:/sw/app/current/#0/database/ip example 127.0.0.1
+kdb meta-set spec:/sw/app/current/#0/database/ip description "ip address of the database server, that the application will connect to"
+kdb meta-set spec:/sw/app/current/#0/database/ip check/ipaddr ''
+```
+
+After this meta-setting bonanza your specification file should look something like this:
+
+```bash
+cat ~/spec.ni
```
Comment (kodebach) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:218]:
```suggestion
cat $(kdb file spec:/sw/app/current/#0)
```

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+```bash
+kdb meta-set spec:/sw/app/current/#0/database/dialect default mssql
+kdb meta-set spec:/sw/app/current/#0/database/dialect example mysql
+kdb meta-set spec:/sw/app/current/#0/database/dialect description "SQL dialect of the database server, that the application will connect to (postgresql|mysql|mssql|mariadb|sqlite)"
+```
+
+After this your specification file should look something like this:
+
+```bash
+cat ~/spec.ni
```
Comment (kodebach) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:274]:
```suggestion
cat $(kdb file spec:/sw/app/current/#0)
```

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+```bash
+kdb meta-set spec:/sw/app/current/#0/backup/date default 2021-11-01
+kdb meta-set spec:/sw/app/current/#0/backup/date example 2021-01-12
+kdb meta-set spec:/sw/app/current/#0/backup/date description "date of the annual server and database backup"
+```
+
+Your specification looks to be complete now! Make sure it look something like the one below and you are good to go using it and configuring the heck out of it!
+
+```bash
+cat ~/spec.ni
```
Comment (kodebach) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:340]:
```suggestion
cat $(kdb file spec:/sw/app/current/#0)
```

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+```
+
+<!-- ------------------------ -->
+## Final specification code
+Duration: 1
+
+After adding all the keys that are necessary for our application to the server, your specification should look something like this:
+
+```bash 
+cat ~/spec.ni
```
Comment (kodebach) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:396]:
```suggestion
cat $(kdb file spec:/sw/app/current/#0)
```

**Review by kodebach — CHANGES_REQUESTED** [review]
In general the tutorial looks really nice. The step-by-step nature will probably help beginners a lot.

There are, however, a few minor things you should change.

It would also be nice, if you could at some point include an example with `kdb set` and `kdb get` to show what the specification actually does. Things like reading a key to see that you automatically get the default value or trying to set a string value as the port, etc.

I would also like to see a note somewhere that the `spec.ni` file can be edited manually in any editor an Elektra will pick up the changes. Otherwise, people might think that you have to use individual `kdb` commands to edit the specification.

**Comment by kodebach** [comment]
> Btw. would be good if you can integrate https://github.com/ElektraInitiative/libelektra/blob/5702ce0496823fdd12c4e1e6ff9d6d1ca1b71a40/tests/shell/shell_recorder/tutorial_wrapper/README.md so that the build server verifies that the tutorial still works.

This probably just means using ` ```sh` instead of ` ```bash` for the code fences.

> Unfortunately, I could not see the result.

It doesn't work perfectly, but you can just open the `index.html` file directly in a Browser and should see what the rendered web page looks like.

@aaronabebe Please also add a quick README that explains how you generated the HTML version from the Markdown file. (The README could also just link to an existing tutorial) Just in case, someone else wants to create a new tutorial or update this one. I would either put this information in `doc/tutorials/interactive/codelabs/README.md` or `doc/tutorials/interactive/codelabs.md`.

**Comment by kodebach** [comment]
A good follow-up to this PR would be to create a build server job (probably Jenkins or Github Actions) that builds the HTML from the Markdown with Google's tool. This job could then be reused for future Codelabs as well.

The job should also publish the results as a website somewhere. I don't think it should be directly integrated into https://libelektra.org (could be complicated and wouldn't look right anyway). Instead we should just add a link to a separate site like we do with https://doc.libelektra.org. We could host the Codelabs on our own server (as e.g. https://codelabs.libelektra.org) or use another service (e.g. Github Pages, for example via [this action](https://github.com/marketplace/actions/deploy-to-github-pages)).

File: doc/tutorials/interactive/codelabs/elektra-specification.md (hunk 1 above)
Comment (markus2330) [reply doc/tutorials/interactive/codelabs/elektra-specification.md:74]: Absolute paths for `spec:` and `system:` will stay absolute and work as expected. (Only for `user:` and `dir:` even absolute paths are not absolute.) As the shell resolves `~` it would be fine to use but I would not recommend to have specs in the home dir. With `pwd` it could be any path. Thus my suggestion.

> sudo kdb mount spec.ni spec:/sw/app/current/#0 ni (possibly with a more unique name instead of spec.ni)

Yes, that would also be okay. But then a person might be surprised where this spec.ni is. (`kdb file` would reveal the answer, though.)

> sudo kdb mount --resolver=noresolver "$PWD/spec.ni" spec:/sw/app/current/#0 ni

I would not recommend "noresolver", it will make storage plugins directly write to config files, which is not safe. "noresolver" was actually designed for non-file based plugins like uname. I created #3823 to clarify that.

File: doc/tutorials/interactive/codelabs/elektra-specification/index.html [hunk 11]
```diff
@@ -0,0 +1,413 @@
...
+      </google-codelab-step>
+    
+      <google-codelab-step label="Summary" duration="1">
+        <ul>
+<li>You setup and mounted a specification using <code>kdb mount</code> and <code>kdb spec-mount</code></li>
+<li>You added keys the specification using <code>kdb meta-set</code></li>
+<li>You added different types of keys with <code>type string</code> or <code>type short</code></li>
+<li>You added keys with enum types, to get specific configuration values, with &#34;</li>
+<li>You added default parameters, examples and descriptions with <code>example</code>, <code>default</code>, <code>description</code></li>
+<li>You also added validation checks using different plugins, like <code>check/port</code>, <code>check/ipaddr</code> or</li>
```
Comment (markus2330) [inline doc/tutorials/interactive/codelabs/elektra-specification/index.html:389]: Sentence stops in the middle.

**Comment by markus2330** [comment]
I now compared the html page rendered by codelabs and the markdown rendered by our website (`mv elektra-specification.md ../.. ` and adding elektra-specification.md in the README.md) and found only two functional differences:

- the time does not count down
- you cannot jump between sections so easily

While I have to admit that the codelabs page looks nice, I am not so sure if these minor differences will pay off to maintain a second place where we have tutorials, especially not if we need to maintain scripts to build these pages.

@aaronabebe For this PR, please fix the things that were found and do `mv elektra-specification.md ../.. ` and add elektra-specification.md in the README.md. I don't think we need to add the html page in the source as it is rendered by the tool?

In general, I like the fresh style of the tutorial and hope we can adapt our other tutorials towards this. Especially the ""What you should already know", What you'll Learn" and "What you'll do" section is something missing in most of the other tutorials.

**Comment by markus2330** [comment]
@aaronabebe for an extra points you can also describe how to use boolean: normalization does not work within the spec, so in the spec 0 and 1 must be used (this definitely should be mentioned in the tutorial) but in the config other values like `true` might be used if you specify the booleans accordingly. Using boolean-specs (as offered by the type plugin) within the spec is an untested area though, so you might hit a bug when investigating in this area.

**Comment by aaronabebe** [comment]
@kodebach I worked in your feedback regarding examples with `kdb get`, `kdb set`  and using vim to directly edit the file. I also added an `README.md` that details how to build the html files from the markdown using the `claat` tool. 

@markus2330  I made a copy of the markdown file at `../../specification.md`, which omits the extra syntax in the markdown that is needed for the google codelab, so that it is just servable as all the other tutorials. I wasn't really sure which would be the final way for it to be in the repository now, thats why I went with copying it. So there is still a possibility for more interactive tutorials and a seperate "interactive hub" in the future. 
Sadly I won't have time to integrate the extra points about boolean normalization you talked about.

**Review by kodebach — COMMENTED** [review]
The changes look very good to me. Just one small thing...

You'll also have to rebase/merge from master, because of a conflict in the release notes.

**Comment by markus2330** [comment]
@aaronabebe The same file is now included two times, once should be enough. Please also make sure to fix the build server issues.

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 12]
```diff
@@ -26,14 +26,14 @@ In this codelab you will learn how to interactively use the `SpecElektra` specif
...
-- use code generation and `kdb` to create and mount a specification for an example CRUD application 
+- use code generation and `kdb` to create and mount a specification for an example CRUD (Create, Read, Update, Delete) application 
 - define defaults, examples and checks for keys in the validation 
 
 <!-- ------------------------ -->
 ## Example App Overview
 Duration: 1
 
-For this tutorial we will write a specification for a very simple CRUD backend application. 
+For this tutorial we will write a specification for a very simple CRUD (Create, Read, Update, Delete) backend application.
```
Comment (markus2330) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:36]: Once should be enough.

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 13]
```diff
@@ -66,27 +66,34 @@ Duration: 5
 ### Step 1: Mount a specification file
 
 First you need to mount a specification file, in this case `spec.ni` to the `spec:/` namespace.
-We define the path inside the `spec:/` namespace as `/sw/app/current/#0`, this is an arbitrary name and you can choose whatever you want here.
+We define the path inside the `spec:/` namespace as `/sw/app/current/\#0`, refer to 
+[the documentation](https://www.libelektra.org/tutorials/integration-of-your-c-application) to find out more about constructing the name.
```
Comment (markus2330) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:70]: Make sure to look at that link, sw names are different (org missing, version/profile needs to be changed)

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 14]
```diff
@@ -194,6 +199,24 @@ server/port =
  meta:/default = 8080
 ```
 
+You can now try to read the value of the newly created configuration. 
+Since you did not set the value to anything yet, we will get the default value back.
+
+```sh
+kdb get /sw/app/current/#0/server/port
+
+8080
```
Comment (markus2330) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:208]: The shell recorder works differently, it needs `#>` to check for values.

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 15]
```diff
@@ -194,6 +199,24 @@ server/port =
...
+```
+
+Try to set the port to `123456` now. 
+
+```sh
+kdb set /sw/app/current/#0/server/port 123456
+```
```
Comment (markus2330) [inline doc/tutorials/interactive/codelabs/elektra-specification.md:215]: You can also check for error codes in the shell recorder.

File: doc/tutorials/interactive/codelabs/elektra-specification.md [hunk 16]
```diff
@@ -204,18 +227,15 @@ Next up you will add a key for the database `ip` address. Like with the key befo
 
 Don't forget the most important rule of configurations: **Always add sensible defaults!**
 
-```bash
```
Comment (markus2330) [inline doc/tutorials/interactive/codelabs/elektra-specification.md]: I would still leave these lines somewhere.

**Review by markus2330 — COMMENTED** [review]
Thank you for the improvements! Please check the build server logs to get this merged. Please remove the duplicate of the file.

**Comment by markus2330** [comment]
>  I made a copy of the markdown file at ../../specification.md, which omits the extra syntax in the markdown that is needed for the google codelab, so that it is just servable as all the other tutorials. I wasn't really sure which would be the final way for it to be in the repository now, thats why I went with copying it. So there is still a possibility for more interactive tutorials and a seperate "interactive hub" in the future.

If you do the removal of codelab-specific-stuff in a separate commit, we can always revert the commit. Then we do not need to have the whole tutorial duplicated in the repo.

> Sadly I won't have time to integrate the extra points about boolean normalization you talked about.

Would be great if you can still do it.

File: doc/tutorials/specification.md [hunk 17]
```diff
@@ -0,0 +1,470 @@
...
+```
+
+Otherwise refer to the [getting started guide](https://www.libelektra.org/getstarted/guide) to install it.
+
+## Mounting the specification
+
+### Step 1: Mount a specification file
+
+First you need to mount a specification file, in this case `spec.ni` to the `spec:/` namespace.
+You can define the path inside the `spec:/` namespace as `/sw/org/app/current/#0`, refer to
```
[the remark this answers is not in the document]
Comment (aaronabebe) [reply doc/tutorials/specification.md]: Oh my bad!

[the remark this answers is not in the document]
Comment (markus2330) [reply doc/tutorials/specification.md]: Thank you for the quick update! Maybe you can also write at least a short paragraph about booleans (that the value 0 and 1 needs to be used in the specification). Or do you plan to further extend anyway?

File: doc/news/_preparation_next_release.md [hunk 18]
```diff
@@ -183,6 +183,14 @@ you up to date with the multi-language support provided by Elektra.
...
 - Improved documentation for module keyset in Elektra Core. _(@lawli3t)_
 - Fixed example in the "mount-configuration-files" tutorial [#3722](https://github.com/ElektraInitiative/libelektra/issues/3722). _(Philipp Oppel)_
+- Reworked [METADATA.ini](/doc/METADATA.ini) _(Markus Raab)_
+- Write notes that `\\` are due to shell recorder, and are not to be copied _(Markus Raab)_
+- Add link to [Go](https://github.com/ElektraInitiative/go-elektra) bindings _(Markus Raab)_
+- Fix order of tutorials _(Markus Raab)_
+- Added API-Reviews for multiple functions in the public API _(Stefan Hanreich)_
+- Minor rewording in [java-kdb.md](/doc/tutorials/java-kdb.md) _(@aaronabebe)_
+- Added interactive and non-interactive versions of a new tutorial to writing a specification _(@aaronabebe)_
+- Add a short Visual Studio 2019 tutorial _(Dominic Jäger)_
```
Comment (mpranj) [inline doc/news/_preparation_next_release.md:193]: Seems like some unrelated lines appeared.
```suggestion
- Added interactive and non-interactive versions of a new tutorial to writing a specification _(@aaronabebe)_
```

Comment (markus2330) [reply doc/news/_preparation_next_release.md]: Thanks for pointing this out! It also looks like doc/news/_preparation_next_release.md now has a merge conflict, so a rebase would be useful anyway.

@aaronabebe you might also want to read doc/GIT.md "Working with forks" and after pulling you do `git rebase master` on your branch.

File: doc/tutorials/specification.md [hunk 19]
```diff
@@ -0,0 +1,518 @@
...
+
+### Step 3: Adding boolean keys
+
+Next up you will configure the `secure` property of our server. This boolean key will toggle if your server encrypts the communication via SSL or not.
+
+So we will add the key and some metadata for it:
+
+```sh
+kdb meta-set spec:/sw/org/app/\#0/current/server/secure type boolean
+kdb meta-set spec:/sw/org/app/\#0/current/server/secure default false
```
Comment (markus2330) [inline doc/tutorials/specification.md:218]: Did you try this? Afaik this does not work, only "0" and "1" work in the spec:

```suggestion
kdb meta-set spec:/sw/org/app/\#0/current/server/secure default 1
```

File: doc/tutorials/specification.md [hunk 20]
```diff
@@ -0,0 +1,518 @@
...
+kdb meta-set spec:/sw/org/app/\#0/current/server/secure example false
```
Comment (markus2330) [inline doc/tutorials/specification.md:219]:
```suggestion
kdb meta-set spec:/sw/org/app/\#0/current/server/secure example 0
```

File: doc/tutorials/specification.md [hunk 21]
```diff
@@ -0,0 +1,518 @@
...
+kdb meta-set spec:/sw/org/app/\#0/current/server/secure description "describes if the REST server uses SSL for communication or not"
+```
+
+By default the `type` plugin will normalize boolean values before storing them.
```
Comment (markus2330) [inline doc/tutorials/specification.md:223]: Afaik, this only works for the config, not for the spec:

**Review by markus2330 — COMMENTED** [review]
Thank you for adding the section about the booleans.

File: doc/news/_preparation_next_release.md (hunk 18 above)
Comment (markus2330) [reply doc/news/_preparation_next_release.md]: Seems like something went wrong with the rebase, now there are even more changes done in other PRs. It is probably best if you cherry-pick your commits in a new branch.

Comment (aaronabebe) [reply doc/news/_preparation_next_release.md]: @markus2330 yeah i don't know what happened. so how would i go about doing that? because i was lazy and just pushed to the `master` of my fork. 

So revert my fork to `upstream/master` and then reapply the commits in a new branch of my fork and redo from that? 
Or should I just open a new PR from a new branch? 
What would be best here? 

Sorry for the inconvinience.

Comment (markus2330) [reply doc/news/_preparation_next_release.md]: Just in case, make a backup of your whole repository, then:

1. Write down the list of commits that are relevant for this PR.
2. Hard reset your master to libelektra master (probably origin/master)
3. git cherry-pick <the list you have from 1.>
4. push --force

**Comment by markus2330** [comment]
@aaronabebe can you please reopen this PR, it would be a pity if this gets lost. Both #3802 and #3893 are outdated.

**Comment by markus2330** [comment]
@aaronabebe any update?

@kodebach can you rebase then please?

**Comment by kodebach** [comment]
Please **do not** reopen this PR. Use #3802 or #3893 instead. This PR used the `master` branch of the `aaronabebe` fork (that's why Github now shows 0 commits in this PR) the other PRs this show commits and changes, because they used separate branches.

@markus2330 Do you have a commit hash of the correct version? Otherwise I can't really rebase it.

**Comment by markus2330** [comment]
> Do you have a commit hash of the correct version?

Unfortunately I couldn't find it. Can you describe for @aaronabebe how he can find it?

**Comment by kodebach** [comment]
> Can you describe for @aaronabebe how he can find it?

Not sure, maybe with `git reflog` in the local repository.

**Comment by aaronabebe** [comment]
@kodebach @markus2330 i will reopen a new PR with the changes from my local repo from a new branch.
Sorry for the delay, june is always a bit of a hassle. I will try to do it this week.

**Comment by markus2330** [comment]
@aaronabebe thank you!
